### ISSUE TITLE
Cheery Cedar Gecko - Allocations in `DonationVotingMerkleDistributionDirectTransferStrategy.sol` or `DonationVotingMerkleDistributionVaultStrategy.sol` can be manipulated to distribute more tokens to specific recepients

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2023-09-Gitcoin-judging#833]
Cheery Cedar Gecko

high

# Allocations in `DonationVotingMerkleDistributionDirectTransferStrategy.sol` or `DonationVotingMerkleDistributionVaultStrategy.sol` can be manipulated to distribute more tokens to specific recepients
## Summary
Allocations in `DonationVotingMerkleDistributionDirectTransferStrategy.sol` or `DonationVotingMerkleDistributionVaultStrategy` can be manipulated to increase the payout of specific recipients in the distribution period.
## Vulnerability Detail
After the registering period in any strategy there will be the allocation phase where things are done different, depending on the strategy used. As can be seen by the Allo documentation,
https://docs.allo.gitcoin.co/strategies/allocation 
the allocation phase for the `Donation voting` strategy involves directly donating to the specific recipients tokens that are allowed. After that in the distribution phase the payouts are proportional to the donations made 
![image](https://github.com/sherlock-audit/2023-09-Gitcoin-VagnerAndrei26/assets/111457602/6948cfb9-6235-4a0b-9ec2-b0fded8dae3f)
which means that if a recipients will get more donations in the allocation phase he will get more payouts in the distribution phase. The problem relies for the cases where `DonationVotingMerkleDistributionDirectTransferStrategy.sol` or `DonationVotingMerkleDistributionVaultStrategy.sol` are allowing any token to be used in the allocation phase by setting `allowedTokens` mapping to `address(0)` 
https://github.com/sherlock-audit/2023-09-Gitcoin/blob/main/allo-v2/contracts/strategies/donation-voting-merkle-base/DonationVotingMerkleDistributionBaseStrategy.sol#L290-L294
When the allocation phase starts any address can allocate to a specific recipient, by donating any amount of tokens, which will first do some checks in `_allocate` 
https://github.com/sherlock-audit/2023-09-Gitcoin/blob/main/allo-v2/contracts/strategies/donation-voting-merkle-base/DonationVotingMerkleDistributionBaseStrategy.sol#L640-L664
and then go trough the `_afterAllocate` hook. In both `DonationVotingMerkleDistributionDirectTransferStrategy.sol` and `DonationVotingMerkleDistributionVaultStrategy.sol` the `_afterAllocate` hook uses `permitTransferFrom` from `Permit2` used by Uniswap to transfer those specific tokens to the recipient himself, in the case of `DonationVotingMerkleDistributionDirectTransferStrategy.sol`
https://github.com/sherlock-audit/2023-09-Gitcoin/blob/main/allo-v2/contracts/strategies/donation-voting-merkle-distribution-direct-transfer/DonationVotingMerkleDistributionDirectTransferStrategy.sol#L59-L72 
or to the contract itself in the case of `DonationVotingMerkleDistributionVaultStrategy.sol`
https://github.com/sherlock-audit/2023-09-Gitcoin/blob/main/allo-v2/contracts/strategies/donation-voting-merkle-distribution-vault/DonationVotingMerkleDistributionVaultStrategy.sol#L121-L131
The problem with using `permitTransferFrom` from `PERMIT2` is the fact that it uses the `SafeTransferLib` from solmate, as can be seen here 
https://github.com/Uniswap/permit2/blob/778de34c6e965f1fea1a82080122a7057c6f24ee/src/SignatureTransfer.sol#L7C9-L7C24
and `SafeTransferLib` from solmate doesn't check if the `token` address has any code at all 
https://github.com/transmissions11/solmate/blob/fadb2e2778adbf01c80275bfb99e5c14969d964b/src/utils/SafeTransferLib.sol#L9
which means that if any token is allowed, anyone can allocate with a token which do not exist, to a specific recipient in very big quantities and the call will succeed. Because of that the whole distribution mechanism for `Donation voting` strategy can be easily manipulated and inflated by doing this multiple time, which will make it so a recipient could get a bigger share of the whole payout, since the payouts are calculated proportional to the donations, as stated in the Allo documentations.
## Code Snippet
https://github.com/sherlock-audit/2023-09-Gitcoin/blob/main/allo-v2/contracts/strategies/donation-voting-merkle-distribution-vault/DonationVotingMerkleDistributionVaultStrategy.sol#L121-L131
https://github.com/sherlock-audit/2023-09-Gitcoin/blob/main/allo-v2/contracts/strategies/donation-voting-merkle-distribution-direct-transfer/DonationVotingMerkleDistributionDirectTransferStrategy.sol#L59-L72
## Tool used

Manual Review

## Recommendation
Do not let every token to be used in the allocation phase, force pool creators to allow only specific tokens for donations or when the allocation phase is happening check for the `codesize` of the token contract intended to be used by the allocators, in that way only token addresses with `codesize` can be used.

**Comment by VagnerAndrei26** [comment]
Escalate
The issue that I have spoken about here it is not the same as the issue that was disputed and excluded more specifically https://github.com/sherlock-audit/2023-09-Gitcoin-judging/issues/922 . The https://github.com/sherlock-audit/2023-09-Gitcoin-judging/issues/922 talks about some checks done and the problem that arise with `_fundPool` function which by sponsor comments, should not be a problem since it enforces some `pool.token`. The issue that I've talked here is that, allocations can be easily manipulated in the cases where all tokens are allowed, more specifically where `allowedTokens[address(0)] = true`. In that case, because the whole allocation mechanic uses Solmate, as specified in the report, allocation with non-existent tokens could happen, which would mess up the whole calculations. There is no real protection for `allocation` like there is for `_fundPool` , as you can see here 
https://github.com/allo-protocol/allo-v2/blob/0b881ef4a0013d2809374c9ea69f4cf1288dfe62/contracts/core/Allo.sol#L352-L354
https://github.com/allo-protocol/allo-v2/blob/0b881ef4a0013d2809374c9ea69f4cf1288dfe62/contracts/core/Allo.sol#L492-L494
so those two strategy that I have talked about can be easily manipulated. I believe this should be treated as a High since the manipulation of allocations would influence also the whole payout mechanism, because as per Gitcoin documentation the payout is proportional to the allocation.

**Comment by sherlock-admin2** [comment]
> Escalate
> The issue that I have spoken about here it is not the same as the issue that was disputed and excluded more specifically https://github.com/sherlock-audit/2023-09-Gitcoin-judging/issues/922 . The https://github.com/sherlock-audit/2023-09-Gitcoin-judging/issues/922 talks about some checks done and the problem that arise with `_fundPool` function which by sponsor comments, should not be a problem since it enforces some `pool.token`. The issue that I've talked here is that, allocations can be easily manipulated in the cases where all tokens are allowed, more specifically where `allowedTokens[address(0)] = true`. In that case, because the whole allocation mechanic uses Solmate, as specified in the report, allocation with non-existent tokens could happen, which would mess up the whole calculations. There is no real protection for `allocation` like there is for `_fundPool` , as you can see here 
> https://github.com/allo-protocol/allo-v2/blob/0b881ef4a0013d2809374c9ea69f4cf1288dfe62/contracts/core/Allo.sol#L352-L354
> https://github.com/allo-protocol/allo-v2/blob/0b881ef4a0013d2809374c9ea69f4cf1288dfe62/contracts/core/Allo.sol#L492-L494
> so those two strategy that I have talked about can be easily manipulated. I believe this should be treated as a High since the manipulation of allocations would influence also the whole payout mechanism, because as per Gitcoin documentation the payout is proportional to the allocation.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by neeksec** [comment]
Suggest to keep invalid.

These strategies are to distribute the donated tokens with the donated amount to target recipient. Not the pool token.

Similar to #784.

**Comment by VagnerAndrei26** [comment]
Yes that is true, but the issue does not talk about the donated tokens themselves, the issue talks about how allocations can be easily inflated right now, because of things that I have speak in my report, which would also lead to inflated payouts in the distribution phase, as per documentation. Also, payouts in the distribution phase are in `pool.token`, not in donations, as you can see per this lines of code 
https://github.com/sherlock-audit/2023-09-Gitcoin/blob/6430c8004017e96ae2f5aac365bdefd0b6eeea72/allo-v2/contracts/strategies/donation-voting-merkle-base/DonationVotingMerkleDistributionBaseStrategy.sol#L774-L800  
which could lead to loss of funds in the end.

**Comment by thelostone-mc** [comment]
> These strategies are to distribute the donated tokens with the donated amount to target recipient. Not the pool token.

+1. We do have allowedTokens for those pools who want to restrict it.
If you use an invalid token -> we'd have to get the USDC value of it off chain , which would equate to 0
So it wouldn't; affect the actual pool token distribution

This becomes a problem when we have the computation done on-chain

**Comment by Evert0x** [comment]
Planning to reject escalation and keep issue state as is

**Comment by Evert0x** [comment]
Result:
Invalid
Duplicate of #922

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [VagnerAndrei26](https://github.com/sherlock-audit/2023-09-Gitcoin-judging/issues/833/#issuecomment-1762848314): rejected
